README.md
# web-diff-files-tree

[![Docker Pulls](https://img.shields.io/docker/pulls/sea5kg/web-diff-files-tree.svg)](https://hub.docker.com/r/sea5kg/web-diff-files-tree/) [![Total alerts](https://img.shields.io/lgtm/alerts/g/sea-kg/web-diff-files-tree.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/sea-kg/web-diff-files-tree/alerts/) [![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/sea-kg/web-diff-files-tree.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/sea-kg/web-diff-files-tree/context:python) [![Language grade: Python](https://img.shields.io/lgtm/grade/cpp/g/sea-kg/web-diff-files-tree.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/sea-kg/web-diff-files-tree/context:cpp) [![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/sea-kg/web-diff-files-tree.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/sea-kg/web-diff-files-tree/context:javascript)

Service which allowed loading filepaths by api. 
After this you can compare different `version` and set comment to files.

![screen](screen.jpg)

[docker](https://hub.docker.com/r/sea5kg/web-diff-files-tree)

## Fast start (via docker-compose)

Just:
```
$ wget -O docker-compose.yml https://raw.githubusercontent.com/sea-kg/web-diff-files-tree/main/docker-compose.yml
$ docker-compose up -d
```

Go look to http://localhost:8071/

for clean:
```
$ docker-compose down
```

Example add data:
```
curl --header "content-type: application/json" \
  --request POST \
  --data '{"group":"product1", "version":"0.0.1", "files"[{"path": "some/file1", "size": 0, "mode": "-rw-rw-r--", "compress_size": 0, "is_dir": false, "dt": "2021-11-18T21:56:08Z"}]}' \
  http://localhost:8071/api/add/
```

## Build (ubuntu)

```
$ sudo apt install \
    cmake \
    git-core \
    pkg-config \
    gcc g++ \
    make \
    libmysqlclient-dev
```

```
./build_simple.sh
```

## Build latest release docker

```
$ docker build --rm --tag sea5kg/web-diff-files-tree:latest -f Dockerfile .
```

## Developing with a docker-compose.develop.yml

```
$ docker-compose -f docker-compose.develop.yml up -d
$ docker exec -it webdifffilestree_develop_server bash
# ./build_simple.sh
# ./web-diff-files-tree
```

Cleanup all:
```
$ sudo rm -rf mysql_data
$ sudo rm -rf tmp
$ docker-compose -f docker-compose.develop.yml down
```


## Database creation (manual)

Create empty database via console client for mysql:

```
$ mysql
> CREATE DATABASE `webdiff` CHARACTER SET utf8 COLLATE utf8_general_ci;
> CREATE USER 'webdiffu'@'localhost' IDENTIFIED BY 'jET3E4W9vm';
> GRANT ALL PRIVILEGES ON webdiff.* TO 'webdiffu'@'localhost' WITH GRANT OPTION;
> FLUSH PRIVILEGES;
```


CHANGELOG.md
# fhq-server Cookbook Changelog

All notable changes to fhq-server project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v0.0.3] - 2021-??-?? (2021 ??? ??)

## [v0.0.2] - 2021-12-16 (2021 Dec 16)

### Changed

- Refoctored code mysql_storage.* to different files
- Redesign page with diff #6
- Improved upload.py

### Added

- Prepared page for Api
- More logs in server log
- Addedshields to README.md

## [v0.0.1] - 2021-12-14 (2021 Dec 14)

### Added

- Init version (redesign from php)


example/upload.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import zipfile
import os
import sys
import re
import stat
import requests
import datetime

# BASEURL = "http://localhost:8071/"
BASEURL = "http://localhost:1234/"

def get_group_and_version(_zip_filename):
    _zip_filename = _zip_filename[:-4] # remove '.zip'
    _items = _zip_filename.split("_")
    match = re.search(r'\d+\.\d+.*',_zip_filename)
    if match is None:
        sys.exit("Could not extract version info from filename")
    version = match.group()
    group = _zip_filename.split(version)[0]
    if group[len(group) - 1] == '_':
        group = group[:-1]
    return (group, version)

def remove_first_folder(_file):
    _file = _file.split("/")
    _file = _file[1:]
    return "/".join(_file)

def upload_files(data_request, _uploaded, _all, _loaded, _all_expected):
    print(" ------- ")
    print("Start adding information. Group: " + data_request['group'] + "; Version: " + data_request['version'])
    print("Adding information about " + str(len(data_request['files'])) + " files ... (done: " + str(_loaded) + "/" + str(_all_expected) + ")")
    r = requests.post(
        BASEURL + 'api/add/',
        json=data_request
    )
    if r.status_code != 200:
        sys.exit("Could not post " + str(r.status_code) + ' ' + str(r.content))
    # print(" " + str(r.json()))
    _uploaded += len(data_request['files'])
    print("Added " + str(_uploaded) + "/" + str(_all_expected))
    return _uploaded

def getfileinfo(_zip, _reg_file):
    ret = {}
    fi = _zip.getinfo(_reg_file)
    ret["path"] = remove_first_folder(_reg_file)
    ret["size"] = fi.file_size
    ret["mode"] = stat.filemode(fi.external_attr >> 16)
    ret["compress_size"] = fi.compress_size
    ret["is_dir"] = fi.is_dir()
    dt = datetime.datetime(
        fi.date_time[0],
        fi.date_time[1],
        fi.date_time[2],
        fi.date_time[3],
        fi.date_time[4],
        fi.date_time[5]
    )
    ret["dt"] = dt.strftime("%Y-%m-%dT%H:%M:%SZ")
    # print(ret)
    return ret

files = os.listdir(".")
for _zip_filename in files:
    if not _zip_filename.endswith(".zip"):
        continue
    data_request = {}
    (group, version) = get_group_and_version(_zip_filename)
    data_request['group'] = group
    data_request['version'] = version
    data_request['files'] = []

    _zip = zipfile.ZipFile(_zip_filename)
    reg_files = _zip.namelist()
    # print(reg_files)
    # for f in reg_files:
    #     # https://docs.python.org/3/library/zipfile.html#zipinfo-objects
    #     fi = _zip.getinfo(f)
    #     filemode = stat.filemode(fi.external_attr >> 16)
    #     print(
    #         f + " size: "
    #         + str(fi.file_size)
    #         + ", compress_size: "
    #         + str(fi.compress_size)
    #         + ", is_dir: "
    #         + str(fi.is_dir())
    #         + ", date_time: "
    #         + str(fi.date_time)
    #         + ", filemode: "
    #         + str(filemode)
    #     )

    _all = len(reg_files)
    _uploaded = 0
    _loaded = 0
    print(" ------- ")
    print("Start adding information. Group: " + data_request['group'] + "; Version: " + data_request['version'])
    all_expected = len(reg_files)
    for _reg_file in reg_files:
        data_request['files'].append(getfileinfo(_zip, _reg_file))
        if len(data_request['files']) > 499:
            _uploaded = upload_files(data_request, _uploaded, _all, _loaded, all_expected)
            _loaded += len(data_request['files'])
            data_request['files'] = []
    if len(data_request['files']) > 0:
        _uploaded = upload_files(data_request, _uploaded, _all, _loaded, all_expected)
        _loaded += len(data_request['files'])
        _uploaded += len(data_request['files'])
        data_request['files'] = []
    print("Done.")
